### DISCUSSION TITLE
API to extract course information from MOOC(s)

### DISCUSSION THREAD

**Discussion by Abhinav281203** [community/community#111208] [Other Feature Feedback, Questions, & Ideas] [answered]
Is there any way to extract course information from all the available MOOC(s) or at least some of them by playing with query parameters or filters? I have been looking for such APIs and encountered various issues while trying out Coursera's and Edx's. Any suggestions? 

Edit: I have completed the task by scraping all of them from Class central. Class central is a website that accumulates courses from various providers. And I forgot to close this thread earlier, Thanks for helping out though.

**Comment by AmauryVanEspen** [comment]
Hi @Abhinav281203,
does this help ? https://docs.openedx.org/projects/edx-platform/en/latest/references/lms_apis.html
Regards
Amaury

**Comment by KuyaMecky** [comment]
Extracting course information from MOOCs can be tricky due to the diversity in platforms and their respective APIs. Here are some suggestions for extracting course information from various MOOC platforms:

1. **Coursera API**: Coursera's public API is limited and not officially documented. You might need to use web scraping techniques to gather information. Be cautious of their terms of service regarding scraping.

2. **edX API**: The edX API is available but has specific endpoints for accessing course information. You may need to authenticate and use their API documentation to fetch the details you need.

3. **Open edX**: If you’re working with an Open edX instance, you can use its REST API. Open edX instances can provide detailed course information through their API endpoints.

4. **Other Platforms**:
   - **Udemy**: Udemy offers a public API that requires authentication. You can access course details and other information through their API.
   - **FutureLearn**: FutureLearn does not provide a public API, so web scraping might be your only option.
   - **Khan Academy**: Khan Academy offers some public APIs, but they are more focused on their own content rather than MOOCs.

5. **Common API Approaches**:
   - **Query Parameters and Filters**: Most APIs allow you to filter results based on parameters like course name, category, or difficulty. Refer to the API documentation for details on how to use these filters effectively.
   - **OAuth and Authentication**: Ensure you are properly authenticated when accessing APIs that require it. This often involves setting up OAuth tokens.

6. **Web Scraping**: For platforms without a public API or if you encounter limitations, you might consider web scraping. Tools like BeautifulSoup (Python) or Puppeteer (Node.js) can be used for scraping. Just be mindful of the terms of service for each site.

7. **Aggregators**: Consider using aggregators or third-party services that compile course information from multiple MOOCs. They might offer APIs or data feeds that you can use.

When using APIs or scraping data, always check the platform's terms of service to ensure compliance with their data usage policies.

**Reply by AmauryVanEspen** [reply]
Hello @KuyaMecky 
do you know if there is some postman collection available for the mentionned API ?
Thanks
Amaury

**Reply by KuyaMecky** [reply]
As of now, there might not be an official Postman collection for the Open edX APIs directly from their documentation. However, you can create your own Postman collection based on the API documentation.

Here’s how you can create a Postman collection for the Open edX APIs:

### Steps to Create a Postman Collection

1. **Set Up Postman**:
   - If you don’t already have Postman, you can download and install it from [Postman's official website](https://www.postman.com/downloads/).

2. **Create a New Collection**:
   - Open Postman and click on "New" > "Collection."
   - Name your collection, e.g., "Open edX API," and add a description if desired.

3. **Add Requests**:
   - For each endpoint you want to use, create a new request in Postman.
   - Set the request type (GET, POST, etc.) and enter the URL based on the Open edX API documentation.

4. **Add Authentication**:
   - If authentication is required, go to the "Authorization" tab in Postman and set up the required authentication method (e.g., Bearer Token).

5. **Set Query Parameters**:
   - For endpoints that require query parameters, use the "Params" tab in Postman to add and configure these parameters.

6. **Save and Test**:
   - Save each request in the collection and test them to ensure they work as expected.

### Sample Requests

- **GET Courses**:
  - Method: GET
  - URL: `https://<your-edx-instance>/api/courses/v1/courses/`
  - Params: Add any query parameters as needed.
  
- **GET Course Details**:
  - Method: GET
  - URL: `https://<your-edx-instance>/api/courses/v1/courses/{course_id}/`
  - Params: Replace `{course_id}` with the actual course ID.

### Existing Collections

You might find unofficial collections or community-made collections on forums or repositories like GitHub. Search for "Open edX Postman collection" on GitHub or similar platforms to see if someone has shared a collection.

If you need help with specific API endpoints or configurations in Postman, let me know!

**Reply by AmauryVanEspen** [reply]
Hi @KuyaMecky 
i'm especially interested in getting answers (and meta, like timestamps) from survey & polls, which are XBlocks in OpenEDX (as far as I know).
It should be related to "problem_responses" endpoints. However, i was not able to succeed. Some ideas ?
Thanks
Amaury

**Reply by KuyaMecky** [reply]
To retrieve survey and poll responses from Open edX, you can use the `problem_responses` API endpoint. This endpoint provides data on student submissions and responses for problem-based XBlocks.

### Key Endpoint
- **GET /api/edx-platform/v1/problems/responses/**

### Parameters
- `course_id`: The ID of the course.
- `user_id`: The ID of the user.
- `block_id`: The ID of the XBlock (survey/poll).

**Example Request:**
```http
GET https://<your-edx-instance>/api/edx-platform/v1/problems/responses/?course_id=course-v1:example+course+2024&user_id=user123&block_id=block-v1:example+course+2024+type@problem+block@survey
```

### Response
The API response will include details such as timestamps, user IDs, and responses. You might need to parse the response to extract specific metadata.

### Troubleshooting
Ensure you have proper authentication and check the endpoint parameters. If the API doesn’t meet your needs, consider direct database queries or creating custom scripts to access XBlock data.

For more detailed exploration, refer to the Open edX API documentation or community forums.

**Reply by AmauryVanEspen** [reply]
Thanks @KuyaMecky I'd got it earlier, unfortunately, it doesn't fit the requirements, to get analytical view (agreagted list of users choices and timestamps by item checked or changes.
Do you know how can i use a wildcard for user_id parameters or how to enhance the current endpoint with enriched answer ?

**Reply by KuyaMecky** [reply]
To get an aggregated view of survey and poll responses from Open edX, you’ll need to extend beyond the basic `problem_responses` endpoint. This endpoint doesn’t support wildcards for `user_id` or aggregation.

You can start by fetching individual responses for each user and then aggregate this data locally using scripts or tools to get metrics like user choices and timestamps. For a more streamlined approach, consider querying the Open edX database directly if you have access, or developing a custom API that provides the aggregated data you need. Checking Open edX forums or community resources may also reveal existing tools or methods for enhanced analytics.

**Comment by Abhinav281203** [comment, marked as answer]
I have completed the task by scraping all of them from Class central. Class central is a website that accumulates courses from various providers. And I forgot to close this thread earlier, Thanks for helping out though.
